gix-fs/src/stack.rs

```diff
@@ -41,15 +46,62 @@ pub trait Delegate {
     fn pop_directory(&mut self);
 }
 
+/// Obtain an iterator over `BStr`-components.
+///
+/// Note that the implementation is simple, and it's mainly meant for statically known strings
+/// or locations obtained during a merge.
+pub trait ToComponents {
+    /// Return an iterator over the components of a path, without the separator.
+    fn to_components(&self) -> impl Iterator<Item = &BStr>;
+}
+
+impl ToComponents for &str {
+    fn to_components(&self) -> impl Iterator<Item = &BStr> {
+        self.split('/').map(Into::into)
+    }
+}
+
+impl ToComponents for String {
+    fn to_components(&self) -> impl Iterator<Item = &BStr> {
+        self.split('/').map(Into::into)
+    }
+}
+
+impl ToComponents for &String {
+    fn to_components(&self) -> impl Iterator<Item = &BStr> {
+        self.split('/').map(Into::into)
+    }
+}
+
+impl ToComponents for BString {
+    fn to_components(&self) -> impl Iterator<Item = &BStr> {
+        self.split(|b| *b == b'/').map(Into::into)
+    }
+}
+
+impl ToComponents for &BString {
+    fn to_components(&self) -> impl Iterator<Item = &BStr> {
+        self.split(|b| *b == b'/').map(Into::into)
+    }
+}
+
+impl ToComponents for &BStr {
+    fn to_components(&self) -> impl Iterator<Item = &BStr> {
+        self.split(|b| *b == b'/').map(Into::into)
+    }
+}
+
 impl Stack {
     /// Create a new instance with `root` being the base for all future paths we handle, assuming it to be valid which includes
     /// symbolic links to be included in it as well.
-    pub fn new(root: PathBuf) -> Self {
+    pub fn new(root: impl ToComponents) -> Self {
```

Comment:
Since this is the root and it is a filesystem location, it must remain a `PathBuf`.

`ToComponents` is really only for the relative paths that are passed in.
It's interesting that `ToComponents` currently doesn't convey information about whether the component is coming from a relative path, which kind of is a problem, but maybe one that can be tackled later.